src/Drips.sol

```diff
@@ -358,14 +266,132 @@ library Drips {
         );
     }
 
+    /// @notice Calculates the end time of drips without duration.
+    /// @param balance The balance when drips have started
+    /// @param dripsStart The timestamp when drips have started.
+    /// @param receivers The list of drips receivers.
+    /// Must be sorted, deduplicated and without 0 amtPerSecs.
+    /// @return defaultEndTime The end time of drips without duration.
+    function _defaultEnd(
+        uint128 balance,
+        uint64 dripsStart,
+        DripsReceiver[] memory receivers
+    ) private pure returns (uint64 defaultEndTime) {
+        require(receivers.length <= MAX_DRIPS_RECEIVERS, "Too many drips receivers");
+        DefaultEndReceiver[] memory defaults = new DefaultEndReceiver[](receivers.length);
+        uint256 length = 0;
+
+        for (uint256 i = 0; i < receivers.length; i++) {
+            DripsReceiver memory receiver = receivers[i];
+            require(receiver.amtPerSec != 0, "Drips receiver amtPerSec is zero");
+            require(i == 0 || _isOrdered(receivers[i - 1], receiver), "Receivers not sorted");
+            uint64 start = receiver.start;
+            if (start < dripsStart) start = dripsStart;
+            if (receiver.duration == 0) {
+                length = _addDefaultEndReceiver(defaults, length, start, receiver.amtPerSec);
```

Comment:
I think we can solve the ordering check very efficient.

As already said, we store the index of the matching index in the first array and remember the highest value from the first array ordering perspective to avoid double entries in the second array.

Reply:
That's a good effort to remove the sorting part of the contract. I'm a little bit worried about the complexity of what you're proposing. From the API perspective it's a redundant parameter which is also difficult to handle by having weird and strict rules about how it should be built. The smart contract logic also expands to optionally support it with verification.

I think that for now we should make it simple and optimize it later, ideally without blowing up the API or the contract.

I've tried coming up with a solution not requiring receivers to be sorted by start time and failed. Finding the default end seems to be logically impossible without an ordered iteration. We've already discussed other API designs and while they wouldn't require sorting, they'd be much less useful.

I've tried different sorting approaches, but they usually boil down to linear searching, which with a hard limit of 100 items isn't unbearable. I've experimented with merge sort, which is theoretically more efficient, but for small lists it turned out to be much less efficient than the naiive algorithms.

Reply:
I think it make sense to get all PRs in. After we have all pieces together, we could still explore potential optimization.

From my perspective it is not such a weird way. We are giving a user the option to pass a specific sorted array instead of building it on-chain.

The verification can be done in O(n) and we anyway need to iterate the array.

However, i agree with the complexity perspective.  I think it is basically a trade-off between gas-optimisation and contract complexity.

I am fine with not implementing it right now. Let's try to get all PR's in. We can still discuss potential optimizations later.